tfhe/src/safe_deserialization.rs

```diff
@@ -0,0 +1,205 @@
+}
+
+/// Deserializes an object with [safe_deserialize] and checks than it is conformant with the given
+/// parameter set
+pub fn safe_deserialize_conformant<T: DeserializeOwned + Named + ParameterSetConformant>(
+    reader: impl std::io::Read,
+    serialized_size_limit: u64,
+    parameter_set: &T::ParameterSet,
+) -> Result<T, String> {
+    let deser: T = safe_deserialize(reader, serialized_size_limit)?;
+
+    if !deser.is_conformant(parameter_set) {
+        return Err(format!(
+            "Deserialized object of type {} not conformant with given parameter set",
+            T::NAME
+        ));
+    }
+
+    Ok(deser)
+}
+
+#[cfg(all(test, feature = "shortint"))]
+mod test_shortint {
+    use crate::safe_deserialization::{safe_deserialize_conformant, safe_serialize};
+    use crate::shortint::parameters::{
+        PARAM_MESSAGE_2_CARRY_2_KS_PBS, PARAM_MESSAGE_3_CARRY_3_KS_PBS,
+    };
+    use crate::shortint::{gen_keys, Ciphertext, PBSParameters};
+
+    #[test]
+    fn safe_ser_ct() {
+        let (ck, _sk) = gen_keys(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+
+        let msg = 2_u64;
+
+        let ct = ck.encrypt(msg);
+
+        let mut buffer = vec![];
+
+        safe_serialize(&ct, &mut buffer, 1 << 40).unwrap();
+
+        assert!(safe_deserialize_conformant::<Ciphertext>(
+            buffer.as_slice(),
+            1 << 40,
+            &PBSParameters::PBS(PARAM_MESSAGE_3_CARRY_3_KS_PBS).to_shortint_conformance_param(),
+        )
+        .is_err());
+
+        let ct2 = safe_deserialize_conformant(
+            buffer.as_slice(),
+            1 << 40,
+            &PBSParameters::PBS(PARAM_MESSAGE_2_CARRY_2_KS_PBS).to_shortint_conformance_param(),
+        )
+        .unwrap();
+
+        let dec = ck.decrypt(&ct2);
+        assert_eq!(msg, dec);
+    }
+}
+
+#[cfg(all(test, feature = "integer"))]
+mod test_integer {
+    use crate::conformance::ParameterSetConformant;
+    use crate::high_level_api::{generate_keys, ConfigBuilder};
+
+    use crate::prelude::{FheDecrypt, FheTryEncrypt};
+    use crate::safe_deserialization::{safe_deserialize_conformant, safe_serialize};
+    use crate::shortint::parameters::{
+        PARAM_MESSAGE_2_CARRY_2_KS_PBS, PARAM_MESSAGE_3_CARRY_3_KS_PBS,
+    };
+    use crate::{CompactFheUint8, CompactPublicKey};
+
+    #[test]
+    fn safe_ser_ct() {
+        let config = ConfigBuilder::all_disabled()
+            .enable_default_integers()
+            .build();
+
+        let (client_key, _server_key) = generate_keys(config);
+
+        let public_key = CompactPublicKey::new(&client_key);
+
+        let msg = 27u8;
+
+        let ct = CompactFheUint8::try_encrypt(msg, &public_key).unwrap();
+
+        assert!(ct.is_conformant(&PARAM_MESSAGE_2_CARRY_2_KS_PBS.into()));
+
+        let mut buffer = vec![];
+
+        safe_serialize(&ct, &mut buffer, 1 << 40).unwrap();
+
+        assert!(safe_deserialize_conformant::<CompactFheUint8>(
+            buffer.as_slice(),
+            1 << 40,
```

Comment:
can we chose a more reasonable size ? 1 << 40 coud exhaust a system memory and people sometimes resort to tests to find reasonable sizes

Reply:
Changed to 1 << 20